### ISSUE TITLE
Unintended heap allocations

### ISSUE THREAD

**Issue by fennel-labs** [Simple-Robotics/proxsuite#91]
ProxQP does a lot of work to avoid heap allocations. However, I found that some heap allocations are still performed after the initialization. Here is the call graph with the functions that allocate from the heap during solve():

![callgraph](https://user-images.githubusercontent.com/12305581/200521991-72c1618c-7021-4f67-91b4-25457b402e8e.png)

## Steps to reproduce:

- Take any QP (e.g. from testcases)
- Compile with:
  - Eigen 3.3.4
  - `-O0 -g -DNDEBUG -std=gnu++17 -DPROXSUITE_VECTORIZE`
- `valgrind --tool=massif --xtree-memory=full <name_to_executable_to_examine>`
- `kcachegrind <name_of_output>`

## Desired behavior
ProxQP does not generate any heap allocations during solve. This is essential for running ProxQP on resource-constrained devices or under real-time constraints.

**Comment by ahoarau** [comment]
Seems to be eigen stuff. It has the nice `Eigen::internal::set_is_malloc_allowed(false);` function to detect runtime allocations.

With https://gitlab.com/libeigen/eigen/-/merge_requests/1055 it even supports reallocs.

**Comment by fennel-labs** [comment]
It is definitely Eigen, as the lowest node of the call graph above already reveals. This means, that Eigen is used wrong or has some unknown version incompatibility with ProxQP.

**Comment by ahoarau** [comment]
Then you can try to surround the solve() call with
```
Eigen::internal::set_is_malloc_allowed(false);
solver->solver();
Eigen::internal::set_is_malloc_allowed(true);
```

NOTE: Don't forget to compile in  `Debug` with EIGEN_RUNTIME_NO_MALLOC defined somewhere.

**Comment by jcarpent** [comment]
@fennel-labs Thanks for raising this issue.
Could you tell us which version of Eigen do you use?

**Comment by fennel-labs** [comment]
I tried with 3.4.0, allocations still happen. Here is the corresponding call graph:

![callgraph_Eigen3 4 0](https://user-images.githubusercontent.com/12305581/200537518-6cf311de-eff4-42a1-b7d1-20797246fc12.png)

**Comment by jcarpent** [comment]
@fennel-labs We have fixed the heap allocations in the code in #92.
Hope it will solve your issue.

**Comment by fennel-labs** [comment]
@jcarpent I can confirm, that this fixes the issue under Eigen 3.4.0. With 3.3.4 heap allocations are still observed.

**Comment by jcarpent** [comment]
Could provide the call graph for 3.3.4?

**Comment by jcarpent** [comment]
@fennel-labs I've just updated the PR with further fixes. Should be fine hopefully.

**Comment by fennel-labs** [comment]
You were a minute faster than me, here is the call graph. I will try the new changes immediately.

![callgraph_Eigen3 3 4_patches](https://user-images.githubusercontent.com/12305581/200579073-5ea6e704-d273-465f-9cf1-b6d5c8dd4dd5.png)

**Comment by fennel-labs** [comment]
@jcarpent Some allocations are still left:
![callgraph_Eigen3 3 4_patches2](https://user-images.githubusercontent.com/12305581/200580569-3b085a50-0d82-4f23-8d0a-96926f8bf663.png)

**Comment by jcarpent** [comment]
@fennel-labs [edba2fa](https://github.com/Simple-Robotics/proxsuite/pull/92/commits/edba2fa92b2c3169b9a855a8002a4165ac62b473) should fix the issue on Eigen 3.3.4

**Comment by fennel-labs** [comment]
@jcarpent Yes, allocations are gone now for both, 3.3.4 and 3.4.0. Thank you so much!

**Comment by jcarpent** [comment]
Thanks @fennel-labs for the quick feedback. We really appreciate it.
